File: functions/src/vendorOpenHours.ts
import {
  FieldValue,
  getFirestore,
  type DocumentData,
  type Query,
  type QueryDocumentSnapshot,
} from "firebase-admin/firestore";
import * as logger from "firebase-functions/logger";
import {onSchedule} from "firebase-functions/v2/scheduler";

/** Sri Lanka — no DST. */
export const COLOMBO_OFFSET_MS = (5 * 60 + 30) * 60 * 1000;

export interface OpeningHours {
  defaultOpen: string;
  defaultClose: string;
  closedSunday: boolean;
}

export interface VendorOpenDoc {
  approvalStatus?: string | null;
  active?: boolean | null;
  openingHours?: unknown;
  openOverrideUntil?: Date | null;
}

export interface SyncVendorOpenResult {
  active: boolean;
  /** `null` means clear the Firestore field. */
  openOverrideUntil: Date | null;
  changed: boolean;
  skippedDueToOverride: boolean;
}

export interface ColomboParts {
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
  second: number;
  dayOfWeek: number;
}

const DEFAULT_OPEN = "09:00";
const DEFAULT_CLOSE = "21:00";

export function toColomboParts(utc: Date): ColomboParts {
  const local = new Date(utc.getTime() + COLOMBO_OFFSET_MS);
  return {
    year: local.getUTCFullYear(),
    month: local.getUTCMonth(),
    day: local.getUTCDate(),
    hour: local.getUTCHours(),
    minute: local.getUTCMinutes(),
    second: local.getUTCSeconds(),
    dayOfWeek: local.getUTCDay(),
  };
}

export function colomboDateTimeToUtc(
  year: number,
  month: number,
  day: number,
  hour: number,
  minute: number,
): Date {
  return new Date(
    Date.UTC(year, month, day, hour, minute, 0, 0) - COLOMBO_OFFSET_MS,
  );
}

export function parseHm(raw: string | null | undefined): {
  hour: number;
  minute: number;
} | null {
  if (typeof raw !== "string") {
    return null;
  }
  const m = /^(\d{1,2}):(\d{2})$/.exec(raw.trim());
  if (!m) {
    return null;
  }
  const hour = Number(m[1]);
  const minute = Number(m[2]);
  if (
    !Number.isInteger(hour) ||
    !Number.isInteger(minute) ||
    hour < 0 ||
    hour > 23 ||
    minute < 0 ||
    minute > 59
  ) {
    return null;
  }
  return {hour, minute};
}

export function parseOpeningHours(raw: unknown): OpeningHours {
  const map =
    raw && typeof raw === "object" ? (raw as Record<string, unknown>) : {};
  const open =
    parseHm(typeof map.defaultOpen === "string" ? map.defaultOpen : null) ?
      String(map.defaultOpen).trim() :
      DEFAULT_OPEN;
  const close =
    parseHm(typeof map.defaultClose === "string" ? map.defaultClose : null) ?
      String(map.defaultClose).trim() :
      DEFAULT_CLOSE;
  return {
    defaultOpen: open,
    defaultClose: close,
    closedSunday: map.closedSunday === true,
  };
}

function minutesOfDay(hour: number, minute: number): number {
  return hour * 60 + minute;
}

function addCalendarDays(parts: ColomboParts, days: number): ColomboParts {
  const noonUtc = colomboDateTimeToUtc(
    parts.year,
    parts.month,
    parts.day,
    12,
    0,
  );
  const shifted = new Date(noonUtc.getTime() + days * 24 * 60 * 60 * 1000);
  return toColomboParts(shifted);
}

/**
 * Whether the shop should accept orders at `now` based on opening hours
 * (Asia/Colombo). Closed all day Sunday when `closedSunday` is set.
 */
export function desiredActive(now: Date, hours: OpeningHours): boolean {
  const parts = toColomboParts(now);
  if (hours.closedSunday && parts.dayOfWeek === 0) {
    return false;
  }
  const open = parseHm(hours.defaultOpen) ?? {hour: 9, minute: 0};
  const close = parseHm(hours.defaultClose) ?? {hour: 21, minute: 0};
  const nowMin = minutesOfDay(parts.hour, parts.minute);
  const openMin = minutesOfDay(open.hour, open.minute);
  const closeMin = minutesOfDay(close.hour, close.minute);

  if (openMin < closeMin) {
    return nowMin >= openMin && nowMin < closeMin;
  }
  // Overnight window, e.g. 22:00 → 02:00.
  return nowMin >= openMin || nowMin < closeMin;
}

/**
 * Next open or close instant after `now` (Asia/Colombo).
 */
export function nextScheduleBoundary(now: Date, hours: OpeningHours): Date {
  const open = parseHm(hours.defaultOpen) ?? {hour: 9, minute: 0};
  const close = parseHm(hours.defaultClose) ?? {hour: 21, minute: 0};
  const openMin = minutesOfDay(open.hour, open.minute);
  const closeMin = minutesOfDay(close.hour, close.minute);
  const overnight = openMin >= closeMin;
  const nowMs = now.getTime();
  const candidates: Date[] = [];
  const start = toColomboParts(now);

  for (let offset = 0; offset <= 8; offset++) {
    const d = addCalendarDays(start, offset);
    const closedDay = hours.closedSunday && d.dayOfWeek === 0;

    if (!overnight) {
      if (closedDay) {
        continue;
      }
      candidates.push(
        colomboDateTimeToUtc(d.year, d.month, d.day, open.hour, open.minute),
      );
      candidates.push(
        colomboDateTimeToUtc(d.year, d.month, d.day, close.hour, close.minute),
      );
      continue;
    }

    if (closedDay) {
      continue;
    }
    candidates.push(
      colomboDateTimeToUtc(d.year, d.month, d.day, open.hour, open.minute),
    );
    const next = addCalendarDays(d, 1);
    if (hours.closedSunday && next.dayOfWeek === 0) {
      candidates.push(
        colomboDateTimeToUtc(next.year, next.month, next.day, 0, 0),
      );
    } else {
      candidates.push(
        colomboDateTimeToUtc(
          next.year,
          next.month,
          next.day,
          close.hour,
          close.minute,
        ),
      );
    }
  }

  candidates.sort((a, b) => a.getTime() - b.getTime());
  for (const c of candidates) {
    if (c.getTime() > nowMs) {
      return c;
    }
  }
  return new Date(nowMs + 24 * 60 * 60 * 1000);
}

export function isApprovalBlocking(approvalStatus: string | null | undefined): boolean {
  return approvalStatus === "pending" || approvalStatus === "rejected";
}

export function syncVendorOpenStatus(
  data: VendorOpenDoc,
  now: Date,
): SyncVendorOpenResult {
  if (isApprovalBlocking(data.approvalStatus ?? null)) {
    const changed =
      data.active !== false || data.openOverrideUntil != null;
    return {
      active: false,
      openOverrideUntil: null,
      changed,
      skippedDueToOverride: false,
    };
  }

  const overrideUntil = data.openOverrideUntil ?? null;
  if (overrideUntil != null && now.getTime() < overrideUntil.getTime()) {
    return {
      active: data.active === true,
      openOverrideUntil: overrideUntil,
      changed: false,
      skippedDueToOverride: true,
    };
  }

  const hours = parseOpeningHours(data.openingHours);
  const desired = desiredActive(now, hours);
  const hadOverride = overrideUntil != null;
  const changed = data.active !== desired || hadOverride;
  return {
    active: desired,
    openOverrideUntil: null,
    changed,
    skippedDueToOverride: false,
  };
}

export function asDate(value: unknown): Date | null {
  if (value == null) {
    return null;
  }
  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? null : value;
  }
  if (typeof value === "object" && value !== null && "toDate" in value) {
    const fn = (value as {toDate: () => Date}).toDate;
    if (typeof fn === "function") {
      const d = fn.call(value);
      return d instanceof Date && !Number.isNaN(d.getTime()) ? d : null;
    }
  }
  if (
    typeof value === "object" &&
    value !== null &&
    "seconds" in value &&
    typeof (value as {seconds: unknown}).seconds === "number"
  ) {
    const stamped = value as {
      seconds: number;
      nanoseconds?: unknown;
      _nanoseconds?: unknown;
    };
    const seconds = stamped.seconds;
    const nanos =
      typeof stamped._nanoseconds === "number" ?
        stamped._nanoseconds :
        typeof stamped.nanoseconds === "number" ?
          stamped.nanoseconds :
          0;
    return new Date(seconds * 1000 + Math.floor(nanos / 1e6));
  }
  return null;
}

export function vendorDocFromFirestore(
  data: DocumentData | undefined,
): VendorOpenDoc {
  if (!data) {
    return {};
  }
  return {
    approvalStatus:
      typeof data.approvalStatus === "string" ? data.approvalStatus : null,
    active: typeof data.active === "boolean" ? data.active : null,
    openingHours: data.openingHours,
    openOverrideUntil: asDate(data.openOverrideUntil),
  };
}

/**
 * Scheduled sync: align `vendors.active` with opening hours unless a manual
 * override is still in effect.
 */
export const syncVendorOpenHours = onSchedule(
  {
    schedule: "every 1 minutes",
    timeZone: "Asia/Colombo",
    region: "asia-south1",
    timeoutSeconds: 120,
    memory: "256MiB",
  },
  async () => {
    const db = getFirestore();
    const now = new Date();
    let updated = 0;
    let scanned = 0;
    let last: QueryDocumentSnapshot | undefined;

    for (;;) {
      let query: Query = db
        .collection("vendors")
        .orderBy("__name__")
        .limit(200);
      if (last) {
        query = query.startAfter(last);
      }
      const snap = await query.get();
      if (snap.empty) {
        break;
      }

      let batch = db.batch();
      let batchCount = 0;

      const commitBatch = async () => {
        if (batchCount === 0) {
          return;
        }
        await batch.commit();
        batch = db.batch();
        batchCount = 0;
      };

      for (const doc of snap.docs) {
        scanned += 1;
        const result = syncVendorOpenStatus(
          vendorDocFromFirestore(doc.data()),
          now,
        );
        if (!result.changed) {
          continue;
        }
        const patch: Record<string, unknown> = {
          active: result.active,
        };
        if (result.openOverrideUntil == null) {
          patch.openOverrideUntil = FieldValue.delete();
        } else {
          patch.openOverrideUntil = result.openOverrideUntil;
        }
        batch.set(doc.ref, patch, {merge: true});
        batchCount += 1;
        updated += 1;
        if (batchCount >= 400) {
          await commitBatch();
        }
      }
      await commitBatch();

      last = snap.docs[snap.docs.length - 1];
      if (snap.size < 200) {
        break;
      }
    }

    logger.info("syncVendorOpenHours complete", {scanned, updated});
  },
);

File: mnd_shop/lib/features/dashboard/domain/vendor_open_hours.dart
/// Asia/Colombo (+05:30, no DST) opening-hours helpers for vendor auto open/close.
///
/// Mirrors `functions/src/vendorOpenHours.ts`.
library;

const Duration kColomboOffset = Duration(hours: 5, minutes: 30);

class OpeningHours {
  const OpeningHours({
    required this.defaultOpen,
    required this.defaultClose,
    required this.closedSunday,
  });

  final String defaultOpen;
  final String defaultClose;
  final bool closedSunday;

  static const OpeningHours defaults = OpeningHours(
    defaultOpen: '09:00',
    defaultClose: '21:00',
    closedSunday: false,
  );
}

class _Hm {
  const _Hm(this.hour, this.minute);
  final int hour;
  final int minute;
}

class _ColomboParts {
  const _ColomboParts({
    required this.year,
    required this.month,
    required this.day,
    required this.hour,
    required this.minute,
    required this.dayOfWeek,
  });

  final int year;
  final int month;
  final int day;
  final int hour;
  final int minute;

  /// `DateTime.sunday` == 7 in Dart; we store 0=Sunday … 6=Saturday.
  final int dayOfWeek;
}

_ColomboParts _toColomboParts(DateTime utc) {
  final DateTime u = utc.toUtc();
  final DateTime local = u.add(kColomboOffset);
  // Treat [local] wall components as Colombo; weekday: Dart Mon=1…Sun=7 → 0=Sun.
  final int dow = local.weekday % 7;
  return _ColomboParts(
    year: local.year,
    month: local.month,
    day: local.day,
    hour: local.hour,
    minute: local.minute,
    dayOfWeek: dow,
  );
}

DateTime _colomboDateTimeToUtc(
  int year,
  int month,
  int day,
  int hour,
  int minute,
) {
  // Interpret wall clock as Colombo, convert to UTC.
  final DateTime asUtcLabel = DateTime.utc(year, month, day, hour, minute);
  return asUtcLabel.subtract(kColomboOffset);
}

_Hm? _parseHm(String? raw) {
  if (raw == null) {
    return null;
  }
  final Match? m = RegExp(r'^(\d{1,2}):(\d{2})$').firstMatch(raw.trim());
  if (m == null) {
    return null;
  }
  final int hour = int.parse(m.group(1)!);
  final int minute = int.parse(m.group(2)!);
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) {
    return null;
  }
  return _Hm(hour, minute);
}

OpeningHours parseOpeningHours(Object? raw) {
  if (raw is! Map) {
    return OpeningHours.defaults;
  }
  final Map<dynamic, dynamic> map = raw;
  final String openRaw = (map['defaultOpen'] as String?)?.trim() ?? '';
  final String closeRaw = (map['defaultClose'] as String?)?.trim() ?? '';
  return OpeningHours(
    defaultOpen: _parseHm(openRaw) != null ? openRaw : OpeningHours.defaults.defaultOpen,
    defaultClose:
        _parseHm(closeRaw) != null ? closeRaw : OpeningHours.defaults.defaultClose,
    closedSunday: map['closedSunday'] == true,
  );
}

int _minutesOfDay(int hour, int minute) => hour * 60 + minute;

_ColomboParts _addCalendarDays(_ColomboParts parts, int days) {
  final DateTime noon = _colomboDateTimeToUtc(
    parts.year,
    parts.month,
    parts.day,
    12,
    0,
  );
  return _toColomboParts(noon.add(Duration(days: days)));
}

/// Whether the shop should accept orders at [now] (Asia/Colombo).
bool desiredActive(DateTime now, OpeningHours hours) {
  final _ColomboParts parts = _toColomboParts(now.toUtc());
  if (hours.closedSunday && parts.dayOfWeek == 0) {
    return false;
  }
  final _Hm open = _parseHm(hours.defaultOpen) ?? const _Hm(9, 0);
  final _Hm close = _parseHm(hours.defaultClose) ?? const _Hm(21, 0);
  final int nowMin = _minutesOfDay(parts.hour, parts.minute);
  final int openMin = _minutesOfDay(open.hour, open.minute);
  final int closeMin = _minutesOfDay(close.hour, close.minute);
  if (openMin < closeMin) {
    return nowMin >= openMin && nowMin < closeMin;
  }
  return nowMin >= openMin || nowMin < closeMin;
}

/// Next open or close instant after [now] (Asia/Colombo).
DateTime nextScheduleBoundary(DateTime now, OpeningHours hours) {
  final _Hm open = _parseHm(hours.defaultOpen) ?? const _Hm(9, 0);
  final _Hm close = _parseHm(hours.defaultClose) ?? const _Hm(21, 0);
  final int openMin = _minutesOfDay(open.hour, open.minute);
  final int closeMin = _minutesOfDay(close.hour, close.minute);
  final bool overnight = openMin >= closeMin;
  final DateTime nowUtc = now.toUtc();
  final int nowMs = nowUtc.millisecondsSinceEpoch;
  final List<DateTime> candidates = <DateTime>[];
  final _ColomboParts start = _toColomboParts(nowUtc);

  for (int offset = 0; offset <= 8; offset++) {
    final _ColomboParts d = _addCalendarDays(start, offset);
    final bool closedDay = hours.closedSunday && d.dayOfWeek == 0;

    if (!overnight) {
      if (closedDay) {
        continue;
      }
      candidates.add(
        _colomboDateTimeToUtc(d.year, d.month, d.day, open.hour, open.minute),
      );
      candidates.add(
        _colomboDateTimeToUtc(d.year, d.month, d.day, close.hour, close.minute),
      );
      continue;
    }

    if (closedDay) {
      continue;
    }
    candidates.add(
      _colomboDateTimeToUtc(d.year, d.month, d.day, open.hour, open.minute),
    );
    final _ColomboParts next = _addCalendarDays(d, 1);
    if (hours.closedSunday && next.dayOfWeek == 0) {
      candidates.add(
        _colomboDateTimeToUtc(next.year, next.month, next.day, 0, 0),
      );
    } else {
      candidates.add(
        _colomboDateTimeToUtc(
          next.year,
          next.month,
          next.day,
          close.hour,
          close.minute,
        ),
      );
    }
  }

  candidates.sort(
    (DateTime a, DateTime b) => a.millisecondsSinceEpoch.compareTo(b.millisecondsSinceEpoch),
  );
  for (final DateTime c in candidates) {
    if (c.millisecondsSinceEpoch > nowMs) {
      return c;
    }
  }
  return nowUtc.add(const Duration(days: 1));
}

bool isApprovalBlocking(String? approvalStatus) {
  return approvalStatus == 'pending' || approvalStatus == 'rejected';
}

class SyncVendorOpenResult {
  const SyncVendorOpenResult({
    required this.active,
    required this.clearOverride,
    required this.changed,
    required this.skippedDueToOverride,
  });

  final bool active;
  final bool clearOverride;
  final bool changed;
  final bool skippedDueToOverride;
}

SyncVendorOpenResult syncVendorOpenStatus({
  required DateTime now,
  required String? approvalStatus,
  required bool? active,
  required Object? openingHours,
  required DateTime? openOverrideUntil,
}) {
  if (isApprovalBlocking(approvalStatus)) {
    return SyncVendorOpenResult(
      active: false,
      clearOverride: openOverrideUntil != null,
      changed: active != false || openOverrideUntil != null,
      skippedDueToOverride: false,
    );
  }

  if (openOverrideUntil != null &&
      now.toUtc().isBefore(openOverrideUntil.toUtc())) {
    return SyncVendorOpenResult(
      active: active == true,
      clearOverride: false,
      changed: false,
      skippedDueToOverride: true,
    );
  }

  final OpeningHours hours = parseOpeningHours(openingHours);
  final bool desired = desiredActive(now, hours);
  final bool hadOverride = openOverrideUntil != null;
  return SyncVendorOpenResult(
    active: desired,
    clearOverride: hadOverride,
    changed: active != desired || hadOverride,
    skippedDueToOverride: false,
  );
}

/// Human-readable time like `9:00 AM` for status subtitles.
String formatColomboClock(DateTime utcOrLocal) {
  final _ColomboParts p = _toColomboParts(utcOrLocal.toUtc());
  final int hour24 = p.hour;
  final int minute = p.minute;
  final String period = hour24 >= 12 ? 'PM' : 'AM';
  int hour12 = hour24 % 12;
  if (hour12 == 0) {
    hour12 = 12;
  }
  final String mm = minute.toString().padLeft(2, '0');
  return '$hour12:$mm $period';
}

enum VendorOpenSubtitleKind {
  awaitingApproval,
  autoOpenUntil,
  autoOpensAt,
  manualOpenUntil,
  manualClosedUntil,
}

class VendorOpenSubtitle {
  const VendorOpenSubtitle({required this.kind, this.clock = ''});

  final VendorOpenSubtitleKind kind;
  final String clock;
}

/// Structured status line under the open/close toggle (localize in UI).
VendorOpenSubtitle vendorOpenStatusSubtitle({
  required bool isOpen,
  required DateTime now,
  required Object? openingHours,
  required DateTime? openOverrideUntil,
  required bool canToggle,
}) {
  if (!canToggle) {
    return const VendorOpenSubtitle(kind: VendorOpenSubtitleKind.awaitingApproval);
  }
  final OpeningHours hours = parseOpeningHours(openingHours);
  final bool overridden = openOverrideUntil != null &&
      now.toUtc().isBefore(openOverrideUntil.toUtc());
  final DateTime until = overridden
      ? openOverrideUntil.toUtc()
      : nextScheduleBoundary(now, hours);
  final String clock = formatColomboClock(until);

  if (overridden) {
    return VendorOpenSubtitle(
      kind: isOpen
          ? VendorOpenSubtitleKind.manualOpenUntil
          : VendorOpenSubtitleKind.manualClosedUntil,
      clock: clock,
    );
  }
  return VendorOpenSubtitle(
    kind: isOpen
        ? VendorOpenSubtitleKind.autoOpenUntil
        : VendorOpenSubtitleKind.autoOpensAt,
    clock: clock,
  );
}

String localizeVendorOpenSubtitle(
  VendorOpenSubtitle subtitle, {
  required String languageCode,
}) {
  final String clock = subtitle.clock;
  final bool si = languageCode == 'si';
  switch (subtitle.kind) {
    case VendorOpenSubtitleKind.awaitingApproval:
      return si
          ? 'Admin අනුමත කළ පසු මෙය සක්‍රීය වේ'
          : 'Available after admin approves your shop';
    case VendorOpenSubtitleKind.autoOpenUntil:
      return si ? 'විවෘතයි $clock දක්වා' : 'Open until $clock';
    case VendorOpenSubtitleKind.autoOpensAt:
      return si ? 'විවෘත වන්නේ $clock' : 'Opens at $clock';
    case VendorOpenSubtitleKind.manualOpenUntil:
      return si
          ? 'අතින් විවෘතයි $clock දක්වා'
          : 'Manually open until $clock';
    case VendorOpenSubtitleKind.manualClosedUntil:
      return si
          ? 'අතින් වසා ඇත $clock දක්වා'
          : 'Manually closed until $clock';
  }
}
